src/tools/gdb/plan_dump.py

```diff
@@ -0,0 +1,610 @@
+		self.__sics = node.cast(ShareInputScan.gdb_type)
+		self.__currentSlice = currentSlice
+
+	def print_node(self):
+		slice_id = -1
+		if self.__currentSlice != 0:
+			slice_id = int(self.__currentSlice["sliceIndex"])
+
+		return "ShareInputScan (share slice:id %d:%d)" % (slice_id, int(self.__sics["share_id"]))
+
+class PartitionSelector(object):
+	gdb_type = gdb.lookup_type("PartitionSelector").pointer()
+
+	def __init__(self, node, reloidMap):
+		self._ps = node.cast(PartitionSelector.gdb_type)
+		self.__reloidMap = reloidMap
+
+	def print_node(self):
+		relname = self.__reloidMap[int(self._ps["relid"])]
+		return "PartitionSelector for %s (dynamic scan id: %d)" % (relname, int(self._ps["scanId"]))
+
+class Scan(object):
+	gdb_type = gdb.lookup_type("Scan").pointer()
+	__typ = ""
+
+	def __init__(self, node, rtableMap, typ):
+		self.scan = node.cast(Scan.gdb_type)
+		self.__rtableMap = rtableMap
+		self.__typ = typ
+
+	def print_node(self, indexid):
+		indexInfo = ""
+		if indexid is not None:
+			indexInfo = " (used index, indexoid %s)" % str(indexid)
+		id = int(self.scan["scanrelid"])
+		relanme = self.__rtableMap[id]
+		return "%s on %s%s" % (self.__typ, relanme, indexInfo)
+
+class PlanDump(gdb.Command):
+	"""Print the plan nodes like pg explain"""
+
+	def __cleanup__(self):
+		self.__result = ""
+		self.__tabCnt = 0
+		self.__curPlanName = None
+		self.__currentSlice = 0
+		self.__pstmt = None
+		self.__state = None
+		self.__rtableMap = {}
+		self.__reloidMap = {}
+
+	def __init__(self):
+		super(PlanDump, self).__init__("plan_dump", gdb.COMMAND_USER, gdb.COMPLETE_SYMBOL)
+		self.__cleanup__()
+
+	def walk_initplans(self, plans, sliceTable):
+		"""
+		This function (and walk_subplans) is an analog of ExplainSubplans
+		from explain.c (for init plans), but instead of working with the PlanStates (like
+		original function does) of plan (the EXPLAIN has all states of
+		plan nodes starting from root), this function can't rely on PlanStates,
+		because coredump's from segments may contain only part planstate
+		(queryDesc->planstate) nodes of the original plan
+		(queryDesc->plannedstmt->planTree), so this function works with the
+		Plan nodes, and it's processing differs (in case of subPlans (not init plans
+		the PlanState nodes have the subPlan field, while the Plan nodes does not))
+		"""
+		head = plans["head"]
+		saved_slice = self.__currentSlice
+		while head != 0:
+			sp = head["data"]["ptr_value"].cast(SubPlan.gdb_type)
+			if sliceTable != 0 and sp["qDispSliceId"] > 0:
+				self.__currentSlice = List.list_nth(sliceTable["slices"], sp["qDispSliceId"]).cast(Slice.gdb_type)
+
+			self.__curPlanName = sp["plan_name"]
+			sp_plan = List.list_nth(self.__pstmt["subplans"], int(sp["plan_id"]) - 1).cast(Plan.gdb_type)
+			self.walk_node(sp_plan)
+			head = head["next"]
+		self.__currentSlice = saved_slice
+
+	def walk_subplans(self, plannode, sliceTable):
+		"""
+		Analog of ExplainSubPlans (only for subplans initPlans are handle by
```

Comment:
handle => handled. Should there be a dot or comma in the brackets: "only for subplans, initPlans"?